packages/block-library/src/table-of-contents/edit.js

```diff
@@ -15,127 +15,228 @@ import {
+	 * they haven't changed (so it returns null). As long as the latest heading
+	 * data remains the same, further calls of the useSelect callback will
+	 * continue to return null, thus preventing any forced re-renders.
+	 */
+	const latestHeadings = useSelect(
+		( select ) => {
 			const {
 				getBlockAttributes,
 				getBlockIndex,
 				getBlockName,
 				getBlockOrder,
+				getBlockParents,
+				getClientIdsOfDescendants,
+				__experimentalGetGlobalBlocksByName: getGlobalBlocksByName,
 			} = select( blockEditorStore );
 
-			const blockIndex = getBlockIndex( clientId );
-			const blockOrder = getBlockOrder();
-
-			// Calculate which page the block will appear in on the front-end by
-			// counting how many <!--nextpage--> tags precede it.
-			// Unfortunately, this implementation only accounts for Page Break and
-			// Classic blocks, so if there are any <!--nextpage--> tags in any
-			// other block, they won't be counted. This will result in the table
-			// of contents showing headings from the wrong page if
-			// onlyIncludeCurrentPage === true. Thankfully, this issue only
-			// affects the editor implementation.
-			let page = 1;
-			for ( let i = 0; i < blockIndex; i++ ) {
-				const blockName = getBlockName( blockOrder[ i ] );
+			// Disable reason: blocks can be loaded into a *non-post* block
+			// editor, so to avoid declaring @wordpress/editor as a dependency, we
+			// must access its store by string. When the store is not available,
+			// editorSelectors will be null, and the block's saved markup will
+			// lack permalinks.
+			// eslint-disable-next-line @wordpress/data-no-store-string-literals
+			const editorSelectors = select( 'core/editor' );
+
+			const pageBreakClientIds = getGlobalBlocksByName( 'core/nextpage' );
+
+			const isPaginated = pageBreakClientIds.length !== 0;
+
+			/**
+			 * Get the relative indices of the block from top to bottom nesting
+			 * level.
+			 *
+			 * @param {string} blockClientId
+			 *
+			 * @return {number[]} The path of indices to the block.
+			 */
+			function getBlockPath( blockClientId ) {
+				const indices = getBlockParents(
+					blockClientId
+				).map( ( ancestorId ) => getBlockIndex( ancestorId ) );
+				indices.push( getBlockIndex( blockClientId ) );
+				return indices;
+			}
+
+			// We can't use just getBlockIndex because it only returns the index relative to sibling blocks, so we have to get all the indices from top to bottom.
+			const tocPath = getBlockPath( clientId );
+
+			// Calculate the page (of a paginated post) this block is part of.
+			// Note that pageBreakClientIds may not be in the order they appear on
+			// the page, so we have to iterate over all of them.
+			let tocPage = 1;
+			for ( const pageBreakClientId of pageBreakClientIds ) {
+				if (
+					comparePathAToB(
+						tocPath,
+						getBlockPath( pageBreakClientId )
+					) < 0
+				) {
+					tocPage++;
+				}
+			}
```

Comment:
Not a blocker here, but it looks like a lot of complexity in this file stems from the unordered data. Starting with sorting, or even better, having some selector options to request ordered data, could shave some comments and a few dozen lines from this file.

Reply:
Thanks to #39985, I was able to simplify the code around here quite a bit.